### DISCUSSION TITLE
fix OpenAPI definition (and thus making all language-specific SDKs obsolete)

### DISCUSSION THREAD

**Discussion by ronreynolds** [smartsheet/smartsheet-java-sdk#94] [Ideas]
been working with a few APIs built from OpenAPI definition files which provide an extreme level of flexibility (language, frameworks, underlying HTTP transport lib, etc) and wondering what's the point of maintaining a language-specific SDK when one can be generated using a build plugin like `org.openapi.generator`?  the "only" requirement is that we fix the OpenAPI definition file at https://smartsheet.redoc.ly/.  unfortunately last i checked that file was totally borked and i know this is a huge task but imagine the benefits if we could actually get it to work!

**Comment by ronreynolds** [comment]
anyway, going to start on this when i have the time; probably start with fixing the sheet endpoints (since those are the ones i use most) and then expand it to the other public endpoints as time/energy allows.  not sure if there exists a public announcement mechanism for up and coming changes to the API but, gee wiz, wouldn't that be something an enterprise-level company like Smartsheet should provide?

**Reply by zromano** [reply]
Hey @ronreynolds, sorry just seeing these discussions. They seem like they will be perfect for communicating.

I think @apederson94 or @DonoA looked at generating our SDK based on the OpenAPI spec  and ran into some challenges. Maybe they can chime in.

**Reply by ronreynolds** [reply]
i know the OpenAPI definition as it exists with our API docs is unusable - huge, heavy with documentation taken straight from Keven F's docs, etc.  so my thinking was to build them by hand (yes, crazy, but at least it would be a starting point) to at least handle the sheet endpoints.  

pretty sure the model is sane (or at least sane enough) because it's Jackson on both ends and i don't remember anything server-side that GENERATED insane JSON (tho i do remember some rather fancy annotations for enum-handling and a few other bits on the receiving side but that's all within the API layer of app-core or the services now handling those endpoints).  so i guess where i would start is the model.  i also believe the API endpoints take quite the variety of URL query-string params (more so, i think, than a "proper REST" API would) but i don't think that would make it un-openapi-able; just somewhat unsavory from an API-purist POV.

**Comment by ronreynolds** [comment]
under way - https://github.com/ronreynolds/smartsheet-openapi
changes so far:
* increased openapi version to 3.1.0 to support DELETE requests with bodies
* converted JSON to YAML for easier reading and editing (never thought i'd say that about YAML)
* removed description and code-samples (code-samples no longer apply to OpenAPI-generated code anyway)
  * guessing the verbosity of the OpenAPI spec is because the docs came first :man_shrugging:

**Comment by DonoA** [comment]
Hey sorry, just seeing this now. So there were a few key challenges I ran in to while attempting to make the swagger generator work on the open api spec:
1. The docs, as you noted, are not really designed for code generation. I found that often times there were comments attached to field type definitions which made them unable to run through the generator. I assume a rewrite will correct this.
2. The API's spec has a few odd cases where the structure of the objects uses a "one of" type model (for example cell values where a number of different objects can be provided with the cell). When I was working on this, these model types were impossible to run through the generator as it doesn't seem to be able to build a jackson config that will deserialize these types.
3. Self referencing types were also un-generatable. I think the workspace endpoints in particular have types which reference them selves or have circular references. The doc generator seems to have no problem with this but the java generator barfed on it.

There were a number of other issues related to the parser itself, however I think a full rewrite will correct these.

**Reply by DonoA** [reply]
I think we can solve these problems if we build iteratively and find where the generator is insufficient, however it was too monumental of a task for me to want to undertake 😬

**Reply by ronreynolds** [reply]
yes, i presumed an iterative approach will be needed too (i.e., remove (for now) those endpoints that can't be supported via OpenAPI and then see if they can be finagled to work with OpenAPI later).  at the moment i'm running into a weird issue with parameter lists that contain multiple refs to other endpoints parameters; the short-term work-around was to replace the refs with a copy of the parameter referenced but hopefully that's not long-term required (tho the only reason i can think of to use refs is to save on disk space and, like any ref, it opens up the possibility of inadvertently breaking one endpoint by changing another so perhaps normalizing all those refs is the best long-term solution (i.e., each endpoint definition is a stand-alone)).

thankfully i have more time now that i'm not doing Smartsheet day-to-day stuff.

**Reply by DonoA** [reply]
> at the moment i'm running into a weird issue with parameter lists that contain multiple refs to other endpoints parameters

If I'm not mistaken this is a bug in the swagger parser I also ran into. Ref tags aren't processed correctly when enclosed in certain parts of the spec. I also fixed this by simply writing a script which copied the content of these files into the parent.

> each endpoint definition is a stand-alone

I wonder how true this is under the hood. A good number of models are shared within grid for each of it's endpoints.

> thankfully i have more time now that i'm not doing Smartsheet day-to-day stuff

I can only dream

**Comment by ronreynolds** [comment]
> I wonder how true this is under the hood. A good number of models are shared within grid for each of its endpoints.

currently dealing with query-string parameters (all of which are, so far, primitives).  not sure one can have multiple schema `$ref`s which should be how TYPES are shared across endpoints (e.g., request body and response payload).  one stack of errors at a time... :)

**Comment by ronreynolds** [comment]
been building a draw.io model of the Etsy OpenAPI spec (for my other work) and have finally circled back to Smartsheet's OpenAPI spec; creating a model for this API would be insane because complex types are defined WITHIN the endpoint definitions (which are then referenced by other endpoint definitions).  no wonder the parser can't handle it.  i also noticed that the ID types (which are longs by nature) are specified as type "number" so need to update all of those.  
just wanted to let folks know i have NOT forgotten about this; i just found it less motivating than i expected to work on Smartsheet code without getting paid.  now i'm starting to work on it just so i don't have my brain atrophy. :) (assuming i can still save it).

**Reply by zromano** [reply]
Haha I would certainly have less motivation if I wasn't being paid by the company. Glad you're still around and willing to help out the SDK though

**Comment by ronreynolds** [comment]
FYI (for those who are interested in OpenAPI specs) there is literally NO difference (according to the generated code) between this:
```
          "createdBy": {
            "description": "User object containing name and email of the creator of this row.",
            "type": "object",
            "allOf": [
              {
                "$ref": "#/components/schemas/NameAndEmail"
              }
            ]
          },
```
and this:
```
          "createdBy": {
            "description": "User object containing name and email of the creator of this row.",
            "$ref": "#/components/schemas/NameAndEmail"
          },
```

**Comment by ronreynolds** [comment]
i noticed
```
          {
            "name": "managedPlanId",
            "in": "query",
            "description": "The target managed plan for which to list events. Authorized if the caller is a system administrator on either the target managed plan or the main plan in EPM hierarchy.",
            "schema": {
              "$ref": "#/components/schemas/Int64"
            },
            "required": false
          }
```
in the parameters of the listEvents endpoint (GET /events) but i don't see it in the `EventResources` class:
```
    EventResult listEvents(Object since, String streamPosition, Integer maxCount, Boolean numericDates) throws SmartsheetException;
```
guessing this is a case where the API has gotten ahead of the SDK?
@zromano ?

**Reply by zromano** [reply]
Yeah, it certainly seems like it. 

We are supposed to have a meeting in the next week or so with leadership to discuss the goals of the SDK and the direction we want to take it

**Reply by ronreynolds** [reply]
well, FWIW, 2 days ago i had a subset of the paths (and all of the schema) parse and compile.  rather than test it i decided to throw the rest of the paths back in so it doesn't parse again but i'm working thru all the `$ref`s now to try to get everything to use schema and parameter elements.  once it parses and compiles again i can start testing it.  pushing everything back to https://github.com/ronreynolds/smartsheet-openapi as i go.  the only piece i ripped out completely were all the `x-codeSamples` since they would be obsolete and are almost certainly dated and not maintained (and simplistic by necessity).

**Reply by ronreynolds** [reply]
still going thru all the `#/paths/` refs (parses but we're back to 500+ compile-time errors).
not sure how the openapi spec was actually generated (certainly it wasn't hand-written) but wondering if a usable openapi-spec json could be generated if the generator was properly configured to favor objects-as-schemas and promoting all reused parameters into the `#/components/parameters/` block since that's basically all i'm doing (just hundreds of times over).
(tagging @zromano in case GitHub doesn't send notifications of replies without tags)

**Reply by DonoA** [reply]
@ronreynolds , I ended up writing a script to relocate and remap all the REF tags when I did this I believe because the import paths were causing substantial issues. IIRC, there are legitimate bugs in how REF tags are handled in the Java Generator project where they get resolved differently based on location.

**Comment by ronreynolds** [comment]
@DonoA  i think i'm starting to find some of the issues you mentioned above; for instance requests like this:
```javascript
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "oneOf": [
                  {
                    "$ref": "#/components/schemas/GroupMember"
                  },
                  {
                    "type": "array",
                    "items": {
                      "$ref": "#/components/schemas/GroupMember"
                    }
                  }
                ]
              }
            }
          }
```
current generate Java code that can't compile; e.g., 
```java
                if (List<GroupMember>.class.equals(Integer.class) || List<GroupMember>.class.equals(Long.class) || List<GroupMember>.class.equals(Float.class) || List<GroupMember>.class.equals(Double.class) || List<GroupMember>.class.equals(Boolean.class) || List<GroupMember>.class.equals(String.class)) {
```
wondering if it's possible to coerce the generator to use actual types rather than `List<Type>` for arrays since getting types out of generics often leads to nasty inner-class trickery that feels like such a hack (and only when it actually works).  :thinking:

**Reply by ronreynolds** [reply]
one thought that comes to mind to address this is that the API server should be able to handle an array of 1 so why make the distinction (or support if it's a PITA) sending both single and an array of elements?  not sure it'll address this but it will simplify things somewhat and there's no reason for a client to support EVERY use-case (at least not use-cases like "supports a single element or an array of 1 or more elements...")

**Reply by ronreynolds** [reply]
this approach is definitely solving the compile-errors (replace `oneOf` of single OR array with just the array).  it should work on both request and response sides so long as the server doesn't generate inconsistent RESPONSES even when given consistent requests. :-/

**Reply by DonoA** [reply]
oneOf was a major source of pain and AFAICT, the java generator just doesn't have great support for this. If possible, I think we should just get rid of our oneOf models, but in lieu of that, I was planning to investigate modifying the spec to use inheritance instead of oneOf. Since the JSON has no class information I assume this would have no impact on the actual request/response objects.

**Comment by apederson94** [comment]
Hey @ronreynolds, I remember that I was going to do some work on figuring out how to replace a lot of the code in our repo with Retrofit. I don't remember who it was, but someone commented on how we have given users of our SDK the ability to pass in their own HTTP clients and how we don't necessarily want to remove that feature. I'm curious on your thoughts about this and how we'd be able to get some of this functionality still with the OpenAPI generated code

**Reply by ronreynolds** [reply]
oh absolutely you can choose the HTTP client and JSON processor to use with openapi-generated code since the spec file is a pure model of the API endpoints, requests, and responses (and a few other bits of behavior).  where the spec becomes real code has gobs of configurability (almost too much but, ironically, never enough.  ROFL).   these are the options with the Java open-api generator - https://openapi-generator.tech/docs/generators/java#config-options
i use the Etsy API with the native client and have to pass in callbacks to get the OAuth headers right so even within a client's framework there's a need for tiny customizations.
not sure if that answers your question, @apederson94 ?

**Reply by apederson94** [reply]
I think this was partially a function of the clients wanting to be able to put their own spin on the HTTP clients at runtime (i.e. they want a different timeout or retry count or whatever). I couldn't seem to parse out whether this is possible with the OpenAPI generated code. Is that something that is still possible?

**Reply by ronreynolds** [reply]
@apederson94 you CHOOSE the http-client library to use when you convert the openapi spec JSON into code.  i'm not sure if openapi-generated code allows for a runtime-pluggable http-client library but i suppose if you generate the code with a http-client that is pluggable it could, in theory, be supported.
as for retry-counts and such that's less a function of the client-library used and more a configuration feature.

**Comment by DonoA** [comment]
@ronreynolds , would having an additional set of hands be helpful with this project? When I was investigating this it was mostly squashing bugs one at a time.

**Reply by ronreynolds** [reply]
@DonoA you can see what i've done thus far at https://github.com/ronreynolds/smartsheet-openapi/tree/work/ron/add-back-paths
unfortunately my latest batch of changes (most of which were to add single-element `"allOf"` arrays so that the descriptions wouldn't get wiped out by the `$ref` sibling).  unfortunately i ran into a 
```
> Task :openApiGenerate FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':openApiGenerate'.
> java.lang.StackOverflowError (no error message)
```
so now i need to pick out the changes until this issue is fixed (the stacktrace doesn't help AFAICT):
```
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':openApiGenerate'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:130)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:282)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:128)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:116)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:74)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:209)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:166)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:42)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:331)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:318)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.lambda$execute$0(DefaultTaskExecutionGraph.java:314)
        at org.gradle.internal.operations.CurrentBuildOperationRef.with(CurrentBuildOperationRef.java:80)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:314)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:303)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:463)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:380)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
Caused by: java.lang.StackOverflowError
        at org.openapitools.codegen.examples.ExampleGenerator.traverseSchemaProperties(ExampleGenerator.java:394)
```
once i get this worked out i'll push the latest openapi spec.  and if i can get past the last 6 build errors (sans the above) i'll merge it into main.

**Reply by DonoA** [reply]
@ronreynolds I can't say if this is related but I ran in to circular reference issues when working with the spec. I believe I was able to use this script to fix it: https://gist.github.com/DonoA/29e3466b5784d4c014f7fe58a0cefb45

Either way it may be helpful as a reference.

**Reply by ronreynolds** [reply]
@DonoA fixed it by rolling back and then reapplying changes incrementally; i think it might have been an empty schema type (that's the only thing i found in the broken version that looked very wrong).  down to 6 compiler errors left; very optimistic that it'll work this week. :)

**Reply by DonoA** [reply]
@ronreynolds , glad to hear it. Sounds like you're making good progress which is impressive because the error messages as I recall and basically worthless. I ended up downloading the source and running it locally to use the debugger to step though for better understanding of the error.

**Reply by ronreynolds** [reply]
@DonoA i actually find the output full of (overflowing) with info but typically the last entries are lacking; more interesting are errors like `Failed to get the schema name: #/paths/~1sheets~1%7BsheetId%7D~1columns/post/requestBody/content/application~1json/schema` and `'oneOf' is intended to include only the additional optional OAS extension discriminator object. For more details, see https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.9.2.1.3 and the OAS section on 'Composition and Inheritance'.` (not sure what to do about the latter error just yet but, as always when debugging, taking out the easy errors first before attacking the potentially complicated ones).

**Comment by ronreynolds** [comment]
@DonoA @apederson94 
ok, it COMPILES.  i had to rip out ALL the "multipart/form-data" responses; will figure out how/if to add them back in later but for now it totally builds and i'll merge that into mainline and tomorrow actually try to use the generated java to, ya know, actually call the Smartsheet API server. :)

**Reply by ronreynolds** [reply]
and i came up with a potential HACK work-around that let me put the responses back in the spec file.  it's not pretty but hacks often aren't.
will push the code back once i get a successful request/response round-trip.  :crossed_fingers: 
also copy-pasting a few bits from the SDK (for instance that rather suspect user-agent-generation code). ;-)

**Reply by ronreynolds** [reply]
the more i think about this the more i suspect these missing-but-referenced types are a bug in the code-generator.  at this rate i might start submitting fixes for that guy too...  :partying_face:

**Comment by ronreynolds** [comment]
invited you all to my openapi project; also here's an MR for your consideration - https://github.com/ronreynolds/smartsheet-openapi/pull/5
next step... get a sheet! :grinning:

**Comment by ronreynolds** [comment]
learned a few new things today - ANYTHING that you want promoted to a type needs to be moved into the components/schemas section (especially if one finds names like "ListSheets200ResponseAllOfDataInner" rather unsavory).  this also includes enums; if they're left in the paths section they're treated as strings with optional validation.  unfortunately we have some enum types (such as sheet-includes and sheet-excludes) that are "enums" with multiple values (normally sent as comma-sep lists of enum values).  in the SDK we have hand-built enums that map to the strings used on the server-side.  thinking will need to change the spec to use arrays for those (tho the to-string bit might be tricky - TBD).

**Reply by ronreynolds** [reply]
hmm.  not exactly a success on the sheet-getting but still progress.
```
[main] INFO com.ronreynolds.smartsheet.api.util.ApiClients - response - (GET https://api.smartsheet.com/2.0/sheets/7290900052922244?one_of_1=0) 200
[main] ERROR com.ronreynolds.smartsheet.BasicUse - failure to get server info
com.ronreynolds.smartsheet.ApiException: com.fasterxml.jackson.databind.JsonMappingException: Unexpected IOException (of type java.io.IOException): Failed deserialization for GetSheet200Response: 2 classes match result, expected 1
	at com.ronreynolds.smartsheet.api.SheetsApi.getSheetWithHttpInfo(SheetsApi.java:892)
	at com.ronreynolds.smartsheet.api.SheetsApi.getSheet(SheetsApi.java:834)
	at com.ronreynolds.smartsheet.api.util.Sheets.getWholeSheet(Sheets.java:126)
	at com.ronreynolds.smartsheet.BasicUse.main(BasicUse.java:17)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Unexpected IOException (of type java.io.IOException): Failed deserialization for GetSheet200Response: 2 classes match result, expected 1
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Unexpected IOException (of type java.io.IOException): Failed deserialization for GetSheet200Response: 2 classes match result, expected 1

	at com.fasterxml.jackson.databind.JsonMappingException.fromUnexpectedIOE(JsonMappingException.java:346)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3852)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3831)
	at com.ronreynolds.smartsheet.api.SheetsApi.getSheetWithHttpInfo(SheetsApi.java:887)
	... 3 more
```
i suspect the spec bits that claim multiple response encodings and different resulting models.  :thinking:

**Reply by DonoA** [reply]
@apederson94 , since I assume these sheet getting enums are processed almost exclusively by GCS, can we simply change the spec to fit what OpenAPI is willing to generate? In this case it sounds like an array of strings instead of a comma separated list.

**Reply by ronreynolds** [reply]
going to switch the single enum to an array of enums and see how that looks (and works); that's technically what it "is" after all.  but basically anything we want to be a TYPE in the generated code needs to be moved into the schema section AFAICT.

**Reply by apederson94** [reply]
@DonoA I think that if it matches the existing API spec then it should work just fine?

**Comment by ronreynolds** [comment]
started a wiki (cuz we all know how much i like to write prose). :wink: 
https://github.com/ronreynolds/smartsheet-openapi/wiki/Smartsheet%E2%80%90OpenAPI
going to merge my current PR in so i can start on the next set of changes.  one really annoying thing about the code is that every single method takes the auth header as the first parameter and given that's embedded in the client config it's 99.999% pointless (unless someone wanted to pass in a different auth header for a specific API call which is so far-left-field they might as well just create a separate `ApiClient`; which reminds me i need to tweak `ApiClients` so one can create one-off `ApiClient` instances that won't replace the current default client.
also should i close out this discussion or is it useful to folks?  it's certainly great to get feedback on this stuff. :grin:

**Comment by ronreynolds** [comment]
kept noticing this weird extra param on my request URLs: `GET https://api.smartsheet.com/2.0/users?one_of_1=0`
turns out that's the "modifiedSince" but because it's defined as an object (of type `DateUnion`) it loses the name assigned in the params and gets encoded as one of the 2 values in the schema:
```
      "DateUnion": {
        "description": "supports date-time in ISO-8601 or epoch-millis",
        "type": "object",
        "oneOf": [
          {
            "type": "string",
            "format": "date-time"
          },
          {
            "$ref": "#/components/schemas/Int64"
          }
        ],
        "readOnly": true
      },
```
not sure how many endpoints actually accept BOTH ISO-8601 date-time strings AND epoch-millis longs but hopefully they don't use a single key for both...

**Reply by ronreynolds** [reply]
that doesn't exactly look like the right format either... :thinking: 
`https://api.smartsheet.com/2.0/users?modifiedSince=2025-02-07T13%3A27%3A03.854447993-08%3A00`
(that's using `OffsetDateTime.now()` as the `modifiedSince`.
i see in `com.smartsheet.api.internal.json.JacksonJsonSerializer` that the `ObjectMapper` is forced to `"yyyy-MM-dd'T'HH:mm:ss'Z'"` which makes sense; wonder if there's a way to do the same with the generated code...

**Reply by ronreynolds** [reply]
:thinking: 
in `ApiClient`:
```
    if (value instanceof OffsetDateTime) {
      return ((OffsetDateTime) value).format(DateTimeFormatter.ISO_OFFSET_DATE_TIME);
    }
```
which is hard-coded (in the JDK) as, e.g., '2011-12-03T10:15:30+01:00'.
this might be a real problem. :-/

**Reply by ronreynolds** [reply]
this seems to hinge on customizations via mustache files ingested by the java code-gen. :exploding_head: 
```
        // helper for client library that allow to parse/format java.time.OffsetDateTime or org.threeten.bp.OffsetDateTime
        if (additionalProperties.containsKey("jsr310") && (isLibrary(WEBCLIENT) || isLibrary(VERTX) || isLibrary(RESTTEMPLATE) || isLibrary(RESTEASY) || isLibrary(MICROPROFILE) || isLibrary(JERSEY2) || isLibrary(JERSEY3) || isLibrary(APACHE) || isLibrary(RESTCLIENT))) {
            supportingFiles.add(new SupportingFile("JavaTimeFormatter.mustache", invokerFolder, "JavaTimeFormatter.java"));
        }
```
still digging into how one of these is written and then how to get code-gen to ingest it.

**Reply by ronreynolds** [reply]
wondering if there's a way to plug in custom ways to convert `List`s into comma-separated strings...
`https://api.smartsheet.com/2.0/sheets/7290900052922244?include=attachments&include=columnType&include=crossSheetReferences&include=discussions&include=filters&include=filterDefinitions&include=format&include=ganttConfig&include=objectValue&include=ownerInfo&include=rowPermalink&include=source&include=writerInfo&rowsModifiedSince=2025-02-07T14%3A46%3A48.728479404-08%3A00) `
i don't think the sheet API currently supports multiple parameters with the same key?

**Reply by ronreynolds** [reply]
the arrays-as-comma-sep-list was pretty easy to solve: https://spec.openapis.org/oas/v3.0.3.html#style-values if you use a style that maps to a collectionFormat=csv then you get just that - a single key=val1,val2,va3...

i also think i figured out why the `JavaTimeFormatter.mustache` isn't getting picked up and it required taking the `openapi-generator-7.11.0.jar` apart (once i found it; didn't know there was a `~/.gradle/...` directory similar to the `~/.m2/...` directory.  anyway, in there are 18 unique library-specific `ApiClient.mustache` files in the Java section; only 9 of them reference `JavaTimeFormatter` and unfortunately the one for `native` does NOT (which is the generated `ApiClient` doesn't extend `JavaTimeFormatter` as expected).  i have no idea why they're all so unique (certainly they differ in ways much more significant than just the HTTP client library being used).  i might be able to work around this by creating our own `ApiClient.mustache` or maybe working with the OpenAPI folks to resolve this (once i figure out how to do it myself).

**Comment by ronreynolds** [comment]
going to try to convert a currently working project to use the openapi-generator code; talk about motivation and testing (and likely a lot of frustration but as long as that leads to improvements it's (hopefully) worth it). :)

**Reply by ronreynolds** [reply]
smartsheet-openapi (v0.1.2) now supports Sheet JSON and i was able to replace the Smartsheet-SDK (and my extension library) with the OpenAPI lib.  feeling very productive. :)
